### ISSUE TITLE
Should the $key property be deleted if it is null or undefined before sending to Firebase?

### ISSUE THREAD

**Issue by mattblang** [angular/angularfire2#197]
For example, when pushing to a FIrebaseListObservable. It would be nice to model Firebase data as classes in TypeScript, but an optional `$key` property will be null when creating a new instance and that causes an error with Firebase.

**Comment by davideast** [comment]
Can you provide a sample of what you're trying to achieve so we're on the same page?

But what I think you're running into is the fact that the Realtime Database won't accept `$` in keys, which is why we use `$key` and `$value`. 

If you want to use a class I would create a method that drops the `$key` from the object and saves it to the database. Look at the example below.

http://jsbin.com/cagudocofa/edit?js,output

```ts
class Item {
  $key?: string;
  name?: string;
  
  /**
  * Create an Item, with no depencies on a DataSnapshot.
  */  
  constructor(key: string, data: Item) {
    this.$key = key;
    this.name = data.name;
  }

  /**
  * Create a reference for the item in the Realtime Database
  */
  ref() {
    return firebase.database().ref().child('items').child(this.$key);
  }

  /**
  * Override the toJSON method to delete the $key
  */
  toJSON() {
    const copy = Object.assign({}, this);
    delete copy.$key;
    return copy;
  }

  /**
  * Write the record in the Realtime Database, without the $key
  */
  save() {
    this.ref().set(this.toJSON());
  }
}

const item = new Item('1', { name: 'itemOne' });
item.save();
```

**Comment by davideast** [comment]
Unless there's any other items for discussion, I'm going to close this issue. Feel free to keep responding though :)

**Comment by mattblang** [comment]
Thanks @davideast for your example! To answer your question, I was wanting to create instances of a model class using the Firebase data that was coming in, and likewise use those instances when updating with the AngularFire2 API. What I noticed is that when I made a new model instance (and the $key was null or undefined depending on how I specified the definition in the constructor), then the Firebase library itself would complain. So, my initial thought had been to delete the $key property if it was null or undefined at the AngularFire2 level to avoid having to do something like override `toJSON`. That change could be bad for reasons I don't realize though.
